File: python/open_guardrail/guards/data_classification.py
"""Classify data sensitivity level."""
from __future__ import annotations

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_LEVELS = ["public", "internal", "confidential", "restricted"]

_PATTERNS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"\btop\s+secret\b", re.IGNORECASE), "restricted"),
    (re.compile(r"\beyes\s+only\b", re.IGNORECASE), "restricted"),
    (re.compile(r"\bclassified\b", re.IGNORECASE), "restricted"),
    (re.compile(r"\brestricted\b", re.IGNORECASE), "restricted"),
    (re.compile(r"\bconfidential\b", re.IGNORECASE), "confidential"),
    (re.compile(r"\bproprietary\b", re.IGNORECASE), "confidential"),
    (re.compile(r"\btrade\s+secret\b", re.IGNORECASE), "confidential"),
    (re.compile(r"\bembargo(ed)?\b", re.IGNORECASE), "confidential"),
    (re.compile(r"\binternal\s+only\b", re.IGNORECASE), "internal"),
]


class _DataClassification:
    def __init__(
        self,
        *,
        action: str = "block",
        min_level: str = "confidential",
    ) -> None:
        self.name = "data-classification"
        self.action = action
        self._min_idx = _LEVELS.index(min_level)

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        highest = "public"
        matched: list[str] = []

        for pat, level in _PATTERNS:
            if pat.search(text):
                matched.append(pat.pattern)
                idx = _LEVELS.index(level)
                if idx > _LEVELS.index(highest):
                    highest = level

        h_idx = _LEVELS.index(highest)
        triggered = h_idx >= self._min_idx
        score = h_idx / (len(_LEVELS) - 1)
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="data-classification",
            passed=not triggered,
            action=(
                self.action if triggered else "allow"
            ),
            score=score,
            message=(
                f"Sensitive data ({highest})"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details={
                "classification_level": highest,
                "matched_patterns": len(matched),
            },
        )


def data_classification(
    *,
    action: str = "block",
    min_level: str = "confidential",
) -> _DataClassification:
    return _DataClassification(
        action=action, min_level=min_level
    )

File: packages/guards/src/data-classification.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

type ClassLevel = 'public' | 'internal' | 'confidential' | 'restricted';

interface DataClassificationOptions {
  action: 'block' | 'warn';
  minLevel?: ClassLevel;
}

const LEVEL_ORDER: ClassLevel[] = ['public', 'internal', 'confidential', 'restricted'];

const CLASSIFICATION_PATTERNS: { pattern: RegExp; level: ClassLevel }[] = [
  { pattern: /\btop\s+secret\b/i, level: 'restricted' },
  { pattern: /\beyes\s+only\b/i, level: 'restricted' },
  { pattern: /\bclassified\b/i, level: 'restricted' },
  { pattern: /\brestricted\b/i, level: 'restricted' },
  { pattern: /\bconfidential\b/i, level: 'confidential' },
  { pattern: /\bproprietary\b/i, level: 'confidential' },
  { pattern: /\btrade\s+secret\b/i, level: 'confidential' },
  { pattern: /\bembargo(ed)?\b/i, level: 'confidential' },
  { pattern: /\binternal\s+only\b/i, level: 'internal' },
];

export function dataClassification(options: DataClassificationOptions): Guard {
  const minLevel = options.minLevel ?? 'confidential';
  const minIdx = LEVEL_ORDER.indexOf(minLevel);

  return {
    name: 'data-classification',
    version: '0.1.0',
    description: 'Classifies data sensitivity level',
    category: 'compliance',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      let highestLevel: ClassLevel = 'public';
      const matched: string[] = [];

      for (const { pattern, level } of CLASSIFICATION_PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
          const lvlIdx = LEVEL_ORDER.indexOf(level);
          if (lvlIdx > LEVEL_ORDER.indexOf(highestLevel)) highestLevel = level;
        }
      }

      const highestIdx = LEVEL_ORDER.indexOf(highestLevel);
      const triggered = highestIdx >= minIdx;
      const score = highestIdx / (LEVEL_ORDER.length - 1);

      return {
        guardName: 'data-classification',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        message: triggered ? `Sensitive data (${highestLevel}) exceeds threshold (${minLevel})` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: { classificationLevel: highestLevel, matchedPatterns: matched.length },
      };
    },
  };
}
